satellite/notifications/chore.go

```diff
@@ -0,0 +1,167 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package notifications
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/rpc"
+	"storj.io/storj/private/sync2"
+	"storj.io/storj/satellite/overlay"
+)
+
+type Chore struct {
+	log     *zap.Logger
+	dialer  rpc.Dialer
+	overlay *overlay.Service
+	Cycle   sync2.Cycle
+}
+
+func NewChore(log *zap.Logger, dialer rpc.Dialer, overlay *overlay.Service, interval time.Duration) *Chore {
+	return &Chore{
+		log:     log,
+		dialer:  dialer,
+		overlay: overlay,
+		Cycle:   *sync2.NewCycle(interval),
+	}
+}
+
+// Run runs notifications report cycle.
+func (chore *Chore) Run(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	return chore.Cycle.Run(ctx,
+		func(ctx context.Context) error {
+			chore.log.Info("sending reports to storagenodes")
+
+			if err := chore.report(ctx); err != nil {
+				chore.log.Error("reporter cycle failed", zap.Error(err))
+			}
+
+			return nil
+		},
+	)
+}
+
+// Close closes underlying cycle.
+func (chore *Chore) Close() (err error) {
+	defer mon.Task()(nil)(&err)
+	chore.Cycle.Close()
+	return nil
+}
+
+// report sends reputation reports to all reliable nodes.
+func (chore *Chore) report(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	sendReports := func(ctx context.Context, nodes []*overlay.NodeDossier) error {
+		for _, node := range nodes {
+			if err = ctx.Err(); err != nil {
+				return err
+			}
+
+			chore.sendReport(ctx, node)
+		}
+
+		return nil
+	}
+
+	return chore.iterate(ctx, sendReports)
+}
+
+// iterate iterates through batches of reliable nodes.
+func (chore *Chore) iterate(ctx context.Context, fn func(context.Context, []*overlay.NodeDossier) error) (err error) {
```

Comment:
Why isn't this part of overlay instead?

Reply:
move it to overlay service is better? Do you need to expose both `db` method and `iterate` or just one of them?

Reply:
I would try to expose as little as possible.

Reply:
well the only reason I did it is for caller to have control over the execution, for example not to abort the loop if some error occured

Reply:
but yeah, can do a function and a wrapper that ignores errors